buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java

```diff
@@ -175,6 +225,158 @@ private void configureNativeVersionGeneration(Project project) {
         );
     }
 
+    private void configureNativeJars(Project project, VariantsExtension variants, boolean testVersionFromLocalRepository) {
+        TaskProvider<Jar> emptyZip = project.getTasks().register("emptyZip", Jar.class, jar -> jar.getArchiveClassifier().set("empty"));
+        // We register the publications here, so they are available when the project is used as a composite build.
+        final Provider<String> artifactId = project.getTasks().named("jar", Jar.class)
+            .flatMap(Jar::getArchiveBaseName);
+        final Provider<Iterable<Jar>> publishableJarTasks = library(project).flatMap(allVariants())
+            .map(toPublishableVariantGroups())
+            .map(toBuildableVariantUsingVariantNames(variants.getVariantNames()))
+            .flatMap(toPublishableJars(project, artifactId));
+
+        project.getConfigurations().named("runtimeElements", configuration -> {
+            final Provider<? extends Iterable<PublishArtifact>> publishableJars = project.getObjects().listProperty(PublishArtifact.class).value(publishableJarTasks.map(it -> stream(it).map(ArchivePublishArtifact::new).collect(toList())));
+            configuration.getOutgoing().getArtifacts().addAllLater(publishableJars);
+        });
```

Comment:
🔍 This interesting side effect of focusing on provider mapping instead of using `afterEvaluate`. Despite the interesting side effect, Gradle still lack some fundamentals to properly use `addAllLater` to effectively defer everything until needed. In this case, we use the internal `ArchivePublishArtifact` type to show how far we could go. Unfortunately, for the publications below, it's not possible.